Make it possible to use WKWebView._historyDelegate when using a non-persistent data store

davidquesada opened the pull request:
#### f5a6af25e43cc53fd17406f1acf8dc804617c15a
<pre>
Make it possible to use WKWebView._historyDelegate when using a non-persistent data store
<a href="https://bugs.webkit.org/show_bug.cgi?id=290326">https://bugs.webkit.org/show_bug.cgi?id=290326</a>
<a href="https://rdar.apple.com/147764328">rdar://147764328</a>

Reviewed by Brady Eidson.

HistoryController currently skips updates that could end up telling the UI process
about history changes if the page is using an ephemeral data store. While this makes
sense in many cases, not all clients want to keep these two decisions (i.e. whether
they have access to this specific set of callbacks about browsing activity, and
whether WebKit should write website data to disk) tightly coupled. Make it possible
to opt a nonpersistent web view in to history callbacks when opting into other
privacy-sensitive operations, i.e. the &quot;AllowPrivacySensitiveOperationsInNonPersistentDataStores&quot;
setting.

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::canRecordHistoryForFrame):
(WebCore::HistoryController::updateForStandardLoad):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::updateForClientRedirect):
(WebCore::HistoryController::updateForSameDocumentNavigation):
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/HistoryDelegate.mm: Added.
(-[HistoryDelegate _webView:didNavigateWithNavigationData:]):
(-[HistoryDelegate _webView:didUpdateHistoryTitle:forURL:]):
(-[HistoryDelegate _webView:didPerformServerRedirectFromURL:toURL:]):
(TEST(HistoryDelegate, NonpersistentDataStoreDoesNotSendHistoryEvents)):
(TEST(HistoryDelegate, NonpersistentDataStoreSendsHistoryEventsWhenAllowingPrivacySensitiveOperations)):
(TEST(HistoryDelegate, PersistentDataStoreSendsHistoryEvents)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UseSystemAppearance.mm:
    Add a missing #import revealed by the shuffling of unified sources as a result
    of adding a new file to TestWebKitAPI.

Canonical link: <a href="https://commits.webkit.org/292628@main">https://commits.webkit.org/292628@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd5cd77d1b328ef5ee66f923fa3d94d003885d5e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96634 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16248 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/6356 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101706 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47154 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98679 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24682 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/73638 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30861 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12442 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87377 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53974 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12201 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5171 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46482 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5263 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103730 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23701 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82689 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23952 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/83426 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82066 "Found 3 new API test failures: /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print-errors, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/print, /WebKitGTK/TestPrinting:/webkit/WebKitPrintOperation/close-after-print (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20606 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26721 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23664 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/28819 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/23323 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25064 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -61,6 +61,21 @@ static inline void addVisitedLink(Page& page, const URL& url)
+    if (!page->usesEphemeralSession())
+        return true;
+
+    if (auto document = frame.document())
```
```suggestion
    if (RefPtr document = frame.document())
```

pxlcoder commented on Source/WebCore/loader/HistoryController.cpp:
```diff
@@ -61,6 +61,21 @@ static inline void addVisitedLink(Page& page, const URL& url)
 
+static inline bool canRecordHistoryForFrame(const LocalFrame& frame)
+{
+    auto page = frame.page();
```
```suggestion
    RefPtr page = frame.page();
```

davidquesada committed:
Make it possible to use WKWebView._historyDelegate when using a non-persistent data store
https://bugs.webkit.org/show_bug.cgi?id=290326
rdar://147764328

Reviewed by Brady Eidson.

HistoryController currently skips updates that could end up telling the UI process
about history changes if the page is using an ephemeral data store. While this makes
sense in many cases, not all clients want to keep these two decisions (i.e. whether
they have access to this specific set of callbacks about browsing activity, and
whether WebKit should write website data to disk) tightly coupled. Make it possible
to opt a nonpersistent web view in to history callbacks when opting into other
privacy-sensitive operations, i.e. the "AllowPrivacySensitiveOperationsInNonPersistentDataStores"
setting.

* Source/WebCore/loader/HistoryController.cpp:
(WebCore::canRecordHistoryForFrame):
(WebCore::HistoryController::updateForStandardLoad):
(WebCore::HistoryController::updateForRedirectWithLockedBackForwardList):
(WebCore::HistoryController::updateForClientRedirect):
(WebCore::HistoryController::updateForSameDocumentNavigation):
(WebCore::HistoryController::pushState):
(WebCore::HistoryController::replaceState):
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/HistoryDelegate.mm: Added.
(-[HistoryDelegate _webView:didNavigateWithNavigationData:]):
(-[HistoryDelegate _webView:didUpdateHistoryTitle:forURL:]):
(-[HistoryDelegate _webView:didPerformServerRedirectFromURL:toURL:]):
(TEST(HistoryDelegate, NonpersistentDataStoreDoesNotSendHistoryEvents)):
(TEST(HistoryDelegate, NonpersistentDataStoreSendsHistoryEventsWhenAllowingPrivacySensitiveOperations)):
(TEST(HistoryDelegate, PersistentDataStoreSendsHistoryEvents)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UseSystemAppearance.mm:
    Add a missing #import revealed by the shuffling of unified sources as a result
    of adding a new file to TestWebKitAPI.

Canonical link: https://commits.webkit.org/292628@main
